src/core/server/status/routes/status.ts

```diff
@@ -0,0 +1,181 @@
+import { PluginName } from '../../plugins';
+import { calculateLegacyStatus, LegacyStatusInfo } from '../legacy_status';
+import { PackageInfo } from '../../config';
+
+const SNAPSHOT_POSTFIX = /-SNAPSHOT$/;
+
+interface Deps {
+  router: IRouter;
+  config: {
+    allowAnonymous: boolean;
+    packageInfo: PackageInfo;
+    serverName: string;
+    uuid: string;
+  };
+  metrics: MetricsServiceSetup;
+  status: {
+    overall$: Observable<ServiceStatus>;
+    core$: Observable<CoreStatus>;
+    plugins$: Observable<Record<PluginName, ServiceStatus>>;
+  };
+}
+
+interface StatusInfo {
+  overall: ServiceStatus;
+  core: CoreStatus;
+  plugins: Record<string, ServiceStatus>;
+}
+
+interface StatusHttpBody {
+  name: string;
+  uuid: string;
+  version: {
+    number: string;
+    build_hash: string;
+    build_number: number;
+    build_snapshot: boolean;
+  };
+  status: StatusInfo | LegacyStatusInfo;
+  metrics: {
+    /** ISO-8601 date string w/o timezone */
+    last_updated: string;
+    collection_interval_in_millis: number;
+    process: {
+      memory: {
+        heap: {
+          total_in_bytes: number;
+          used_in_bytes: number;
+          size_limit: number;
+        };
+        resident_set_size_in_bytes: number;
+      };
+      event_loop_delay: number;
+      pid: number;
+      uptime_in_millis: number;
+    };
+    os: {
+      load: Record<string, number>;
+      memory: {
+        total_in_bytes: number;
+        used_in_bytes: number;
+        free_in_bytes: number;
+      };
+      uptime_in_millis: number;
+      platform: string;
+      platformRelease: string;
+    };
+    response_times: {
+      max_in_millis: number;
+    };
+    requests: {
+      total: number;
+      disconnects: number;
+      statusCodes: Record<number, number>;
+      status_codes: Record<number, number>;
+    };
+    concurrent_connections: number;
+  };
+}
+
+export const registerStatusRoute = ({ router, config, metrics, status }: Deps) => {
+  // Since this observable is not subscribed to elsewhere, we need to subscribe
+  // here to eagerly load the plugins status when Kibana starts up.
+  const plugins$ = new ReplaySubject<Record<PluginName, ServiceStatus>>();
+  status.plugins$.subscribe(plugins$);
+
+  router.get(
+    {
+      path: '/api/status',
+      options: {
+        authRequired: !config.allowAnonymous,
+        tags: ['api'], // ensures that unauthenticated calls receive a 401 rather than a 302 redirect to login page
+      },
+      validate: {
+        query: schema.object({
+          v8format: schema.boolean({ defaultValue: false }),
+        }),
+      },
+    },
+    async (context, req, res) => {
+      const { version, buildSha, buildNum } = config.packageInfo;
+      const versionWithoutSnapshot = version.replace(SNAPSHOT_POSTFIX, '');
+      const [overall, core, plugins] = await combineLatest([
+        status.overall$,
+        status.core$,
+        plugins$,
+      ])
+        .pipe(first())
```

Comment:
Unfortunately, without a `subscribe` call it seems that the Observable will not start until the very first API call to `/api/status`, even with a shareReplay. This does make sense since Observables are designed to be a pull-based mechanism, but I also wonder if there isn't some feature or utility that I'm missing here.

Using the manual ReplaySubject + subscribe decreases the first response time from 8500ms (!!) down to about 5ms. It's a bit hacky, but I think it could be critical on this endpoint since this serves as our health check to external systems.

If there's some other RxJS-fu that can be done, I'm all ears!

Reply:
I do wonder if we should add a timeout on the plugin-supplied status Observables so that a single plugin's status cannot block this endpoint 🤔

Reply:
> I do wonder if we should add a timeout on the plugin-supplied status Observables so that a single plugin's status cannot block this endpoint

That makes sense to me, but doing it as a follow-up is fine to me